pkg/server/batch_check.go

```diff
@@ -0,0 +1,131 @@
+package server
+
+import (
+	"context"
+	"errors"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/openfga/openfga/pkg/middleware/validator"
+
+	"github.com/openfga/openfga/internal/condition"
+	"github.com/openfga/openfga/internal/graph"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	"github.com/openfga/openfga/internal/authz"
+	"github.com/openfga/openfga/pkg/server/commands"
+	"github.com/openfga/openfga/pkg/telemetry"
+)
+
+func (s *Server) BatchCheck(ctx context.Context, req *openfgav1.BatchCheckRequest) (*openfgav1.BatchCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, authz.BatchCheck, trace.WithAttributes(
+		attribute.KeyValue{Key: "store_id", Value: attribute.StringValue(req.GetStoreId())},
+		attribute.KeyValue{Key: "batch_size", Value: attribute.IntValue(len(req.GetChecks()))},
+		attribute.KeyValue{Key: "consistency", Value: attribute.StringValue(req.GetConsistency().String())},
+	))
+	defer span.End()
+
+	if !validator.RequestIsValidatedFromContext(ctx) {
+		if err := req.Validate(); err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+
+	err := s.checkAuthz(ctx, req.GetStoreId(), authz.BatchCheck)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = telemetry.ContextWithRPCInfo(ctx, telemetry.RPCInfo{
+		Service: s.serviceName,
+		Method:  authz.BatchCheck,
+	})
+
+	storeID := req.GetStoreId()
+
+	typesys, err := s.resolveTypesystem(ctx, storeID, req.GetAuthorizationModelId())
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := commands.NewBatchCheckCommand(
+		s.checkDatastore,
+		s.checkResolver,
+		typesys,
+		commands.WithBatchCheckCommandCacheController(s.cacheController),
+		commands.WithBatchCheckCommandLogger(s.logger),
+		commands.WithBatchCheckMaxChecksPerBatch(s.maxChecksPerBatchCheck),
+		commands.WithBatchCheckMaxConcurrentChecks(s.maxConcurrentChecksPerBatch),
+	)
+
+	result, err := cmd.Execute(ctx, &commands.BatchCheckCommandParams{
+		AuthorizationModelID: req.GetAuthorizationModelId(),
+		Checks:               req.GetChecks(),
+		Consistency:          req.GetConsistency(),
+		StoreID:              storeID,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &openfgav1.BatchCheckResponse{Result: transformCheckResultToRPC(result)}, nil
+}
+
+// transformCheckResultToRPC takes about 150ns per check, or 0.00015ms.
+// If batch sizes ever increase into the thousands, we may want to
+// refactor this to work concurrently alongside the checks themselves.
```

Comment:
**Request 1:**  I definitely see why you make this note butt he doc comment shouldn't include these performance notes. Instead, should describe the function. 

**Request 2:**  Because performance is a concern here, can you add a benchmark test for it? Also it looks like it needs unit tests too?